### PR TITLE
SD card detection + more

### PR DESCRIPTION
**Pull request by nahuhh**
IMPORTANT NOTES FOR PRE-2022 **INTERNAL STORAGE** USERS:
1. Select "N” when asked to setup termux storage
2. Select "No" when asked if using an SD card
Script will check for storage space and move the blockchain to the appropriate location (crypto/monero-cli on internal storage)
3. If not enough space, script will continue with things in place. 

All other users (new or prior SD card users):
1. Proceed as usual.

Release Notes
Features:
- changed config folder to public storage
- blockchain on internal will be stored next to config, in "crypto/monero-cli" folder.
- will auto prune if using internal storage and < 150gb available
- check for SD card (hackish) to work for non-rooted devices. This works by checking for storage/external-1 in termux.
- create symlinks in termux' monero-cli folder to config and blockchain data

Fixes:
- remove forgotten blocklist download

### CODE REVIEW

**Comment by CryptoGrampy** [comment]
Really excited to test this out- should get around to it tomorrow.

File: src/full-monero-node-external-sd-install.sh [hunk 1]
```diff
@@ -3,8 +3,9 @@
 (
 MONERO=~/monero-cli
 MONERO_CLI=~/monero-cli/monero-cli
-NODE_DATA=~/storage/external-1/bitmonero
-NODE_CONFIG=~/monero-cli/config
+NODE_CONFIG=~/storage/shared/crypto/monero-cli/config
+INTERNAL_NODE=~/storage/shared/crypto/monero-cli/blockchain
```
Comment (CryptoGrampy) [inline src/full-monero-node-external-sd-install.sh:7]: how about `INTERNAL_NODE=~/storage/shared/crypto/monero-cli/bitmonero`

File: src/full-monero-node-external-sd-install.sh [hunk 2]
```diff
@@ -43,23 +45,104 @@ termux-wake-lock -y
...
+mkdir -p $NODE_CONFIG
 mkdir -p $TERMUX_BOOT
 mkdir -p $TERMUX_SHORTCUTS
 mkdir -p $TERMUX_SCHEDULED
-mkdir -p $NODE_CONFIG
 
+# Pre-Clean Old Setup
+rm -f $TERMUX_BOOT/before_start_monero_node
+rm -f $TERMUX_BOOT/*XMR\ Node*
+NODE_CONFIGOLD=~/monero-cli/config
```
Comment (CryptoGrampy) [inline src/full-monero-node-external-sd-install.sh]: `NODE_CONFIG_OLD`

File: src/full-monero-node-external-sd-install.sh [hunk 3]
```diff
@@ -43,23 +45,104 @@ termux-wake-lock -y
...
+if [ -d $NODE_CONFIGOLD ]
+then
+mv $NODE_CONFIGOLD/* $NODE_CONFIG/
+rm -r $NODE_CONFIGOLD
+fi
 
-# Download Blocklist (disabled)
-cd $NODE_CONFIG
-# wget -O block.txt https://gui.xmr.pm/files/block.txt
+SD=~/storage/external-1
```
Comment (CryptoGrampy) [inline src/full-monero-node-external-sd-install.sh:71]: Good spot for a comment denoting what's happening in this section... i.e. something like `# Set blockchain location - internal or external SD`

**Comment by CryptoGrampy** [comment]
First of all, this looks amazing.  My only main comments (before I start testing): 
1. Keep variable names consistent with `WORDS_CONSISTENTLY_SEPARATED`
2. Have you looked into: https://github.com/monero-project/monero-gui/issues/3619 ? Thoughts? It might be harmful to internal storage to hold the blockchain.  It's definitely worth adding a warning, but totally worth testing out to see how long a device lasts.
3. Let's make sure there aren't any weird storage limits with `shared`.  I could have sworn there were some kind of app limitations on how much could be stored there (though you can store a ton of camera images there) that I saw during the initial development of this, but let's just make sure there aren't any issues.  Feel free to let me know if you've already tested this.

**Comment by nahuhh** [comment]
> First of all, this looks amazing.  My only main comments (before I start testing): 
> 1. Keep variable names consistent with `WORDS_CONSISTENTLY_SEPARATED`

Underscores are annoying 😂, but yeah, go ahead and fix them up if you prefer.

I noted the instances where I miss the underscore on the note. You can probably do a quick search and replace. I'm not using a pc.

You can the names too.. I wrote them in a way that I, personally, would remember when testing. 

> 2. Have you looked into: https://github.com/monero-project/monero-gui/issues/3619 ? Thoughts? It might be harmful to internal storage to hold the blockchain.  It's definitely worth adding a warning, but totally worth testing out to see how long a device lasts.

I believe the idea that internal is bad is unrealistic. Internal storage is usually higher quality than micro SD cards. 
The issue would, in theory, be due to the amount/frequency of writes, assuming 0 downtime. 

But even still: According to below, a test would take ~50 years. (There may be a difference / might be worse for pruned if pruning does more rewrites. But I assume still 25 years on paper, and at least 5-10 years in practice on high quality internal storage.)

Node on my end is doing ~3.5gb/day downloads, after fully synced. 


Writes on a phone are estimated at  ~10gb/day 


Android auto is estimated to be higher. 
It's noted that the larger the storage space, the longer the lifespan. But assuming you are writing an amount equal to to the full storage space per day (16gb/day on 16gb storage - so 100% rewrite per day) lifespan for 3k writes lifespan, is ~ 10 years. 

Source: [Android.com](https://source.android.com/devices/tech/perf/flash-wear)

On internal storage of 128gb, a pruned may write the equivalent of the full blockchain (ie, greater than 100% of the storage space), but this only happens once. Aprox 1/3000 rewrites. 

+ Internal storage should be better than a pc with an hdd. But iirc, docs don't have warnings for hdd re longevity (?) but do have warnings for speed 
+ docs recommends ssd, again for speed. SSD is a form of flash storage, again, no longevity warnings (?)
+ there is a notice in the readme - though, I'm not sure if needed (?)

(?) = request for comment. I may be incorrect 

> 3. Let's make sure there aren't any weird storage limits with `shared`.  I could have sworn there were some kind of app limitations on how much could be stored there (though you can store a ton of camera images there) that I saw during the initial development of this, but let's just make sure there aren't any issues.  Feel free to let me know if you've already tested this.  
> 

I haven't synced enough to "shared".
Well, I may have, but don't remember. 
I'll sync to 5gb or so and post an update

Update:
Synced 5gb np.


Downloaded 2.23gb


Actual storage used 5gb

Disclaimer
Don't shoot me if I'm dead wrong

File: src/full-monero-node-external-sd-install.sh (hunk 1 above)
Comment (nahuhh) [reply src/full-monero-node-external-sd-install.sh:7]: Haven uses .haven
Wownero uses .wownero
I'm assuming 
When monero was bitmonero, it used .bitmonero
And that
Monero uses .bitmonero, for backwards compatibility. 

Since we're creating a new folder to store the blockchain, as the default folder is inside of termux app directory, I dont like the pattern

Ie
Monero blockchain folder would be monero

I think "blockchain" is fine.. It's next to "config".
Bitmonero is an old name and "monero" doesn't imply that the blockchain is stored there.

File: src/full-monero-node-external-sd-install.sh (hunk 3 above)
Comment (nahuhh) [reply src/full-monero-node-external-sd-install.sh:71]: The comment is 2 lines below. Feel free to move

File: src/full-monero-node-external-sd-install.sh (hunk 2 above)
Comment (nahuhh) [reply src/full-monero-node-external-sd-install.sh]: There's a bunch of these (with no underscore separating the words).

Is this just aesthetic?  

ASKSD
INTFREE
MOVEDATA
CONFIRMINT
NODE_CONFIGOLD

**Comment by CryptoGrampy** [comment]
- Awesome research on the internal longevity... I think the best approach is to simply prompt and tell the user that this COULD shorten the life of their phone's internal storage.  Proceed? Y/N
- Variable names- yeah I honestly don't really care a ton and may just go back and tighten them up later.  Really not a big deal.


- As for running this, I think I'm having some permissions issues when using the SD install with update or keep config: 
`mkdir: cannot create directory ‘/data/data/com.termux/files/home/storage/shared’: Permission denied
mv: target '/data/data/com.termux/files/home/storage/shared/crypto/monero-cli/config/' is not a directory
Using SD Card
sh: 141: cd: can't cd to monero
ln: failed to create symbolic link 'blockchain': File exists
ln: failed to create symbolic link 'config': File exists
sh: 145: cd: can't cd to /data/data/com.termux/files/home/storage/shared/crypto/monero-cli/config
Keep existing config file.`
Simply re-running the latest master branch script brings things back to normal for me.

**Comment by nahuhh** [comment]
I'll update when ready to test again.

> `mkdir: cannot create directory ‘/data/data/com.termux/files/home/storage/shared’: Permission denied

Can you test: 

Are you able to manually CD into shared?
```
cd storage/shared/ && ls
```
If so, can you create "mkdir" or "touch" there? 
If so, I can fix this.
Probably just have to CD into shared to make the directories. 

Btw. There isn't anywhere in the script that "mkdir -p ~/storage/shared" . Onky ~storage/shared/crypto/* 
So I think this error is coming from the -p flag (make parents).

If you can cd into shared, shoukd confirm. If not, perhaps I need to use storage/downloads for compatibility. 

Fwiw, I'm testing on 
SD card 
Android 7 and 11

Internal
Android 10

> sh: 141: cd: can't cd to monero

My mistake. (Fixed)
(As a ressult of the bad CD, the script likely created your "blockchain" and "config" symlinks in the termux home directory. Are they? If so, the below command worked)

> ln: failed to create symbolic link 'blockchain': File exists
> ln: failed to create symbolic link 'config': File exists

I had ln running without the -f flag. (I added the flag)

**Comment by CryptoGrampy** [comment]
Okay, so for whatever reason, I was *not* able to cd into shared at all, so I deleted Termux, reinstalled, installed the latest master (it worked fine), then installed using your latest commit, and everything worked great and I could cd into shared so :shrug: .  

Also want to confirm that the install works great on internal storage. 

Edit: Okay maybe some things aren't working- starting and stopping from the widget launcher seems to not be working correctly or stops working after a while - researching now...  This is happening on a phone running with SD as well as phone running internally.

I seem to be losing access to the shared folder, or perhaps weird things are happening with symlink.  Not able to really dig in a ton at the moment: https://www.reddit.com/r/termux/comments/o98ulg/keep_losing_permission/

File: src/full-monero-node-external-sd-install.sh (hunk 3 above)
Comment (nahuhh) [reply src/full-monero-node-external-sd-install.sh:71]: Moved

**Comment by nahuhh** [comment]
> Okay, so for whatever reason, I was *not* able to cd into shared at all, so I deleted Termux, reinstalled, installed the latest master (it worked fine), then installed using your latest commit, and everything worked great and I could cd into shared so :shrug: .  
> 
> Also want to confirm that the install works great on internal storage. 
> 
> Edit: Okay maybe some things aren't working- starting and stopping from the widget launcher seems to not be working correctly or stops working after a while - researching now...  This is happening on a phone running with SD as well as phone running internally.
> 
> I seem to be losing access to the shared folder, or perhaps weird things are happening with symlink.  Not able to really dig in a ton at the moment: https://www.reddit.com/r/termux/comments/o98ulg/keep_losing_permission/

Yeah, looks to be random/widespread. 
Been going on for a few years it seems. 

I figure theres not much we can do about it - people losing storage permission on termux, that is. 

If it works, it works. 
There are a few devices that arent working for me at all
- A Huawei p20 (Segmentation fault. Internal storage)
- An android s905x2 android tv (bus error. Internal and sd)

I don't have the time to diagnose those at the moment, but I'm assume there more devices that are incompatible out of the box. 


TLDR

- Termux workaround is to re enable storage permission. 
- If user cannot use an sd card, and is losing access repeatedly.. they are SOL and should use a different device or an sd card.
- If it doesn't work at all (bus or seg fault).. Again, SOL. Find a new device and try again. 

Have you thought about adding buttons to the running notification to stop & check status? 

Latest commit should be good to go. Let me know if you find anything out of place

**Comment by nahuhh** [comment]
@plowsof 
Want to give it a run?

@CryptoGrampy 
Are you still losing storage access? 
Can you reproduce how to lose it? 
Couple different things to try after confirming access:
- force close (losing access on exit) 
- reboot (losing access on reboot)
- run termux storage setup, then reboot (overwriting settings error back to default)
- mkdir in shared, then reboot (revoking after usage) 
Or is it completely random?

Also, after losing access:
- can you still access or use storage/downloads 
- are files you created there accessible from a file manager? 
If so, perhaps I can move from shared/crypto to downloads/crypto

(Will pose an issue for the config file if access is lost as config is stored in shared/crypto for SD card users as well)

**Comment by CryptoGrampy** [comment]
I should also be able to give it another whirl today

**Comment by plowsof** [comment]
@nahuhh Apologies for the slow response, i've been swamped with another project, i will give this a whirl as soon as i can :+1:

**Comment by nahuhh** [comment]
https://github.com/nahuhh/android-termux-monero-node/releases/tag/V4.0.2

sh -c "$(curl -fsSL https://github.com/nahuhh/android-termux-monero-node/raw/4.0.2/src/install-monerod-in-termux.sh)" 

New version has notification updates 
- release / acquire wakelock from the notification
- "XMR Node Status" doesn't refresh the notification anymore. Just queries the node for print_net_stats and status. 
(You can refresh the notification from the notification)
